### PR TITLE
propagation of hcal geometry changes for the csc halo algo

### PR DESCRIPTION
**Pull request by zdemirag**
Related PRs are:
* https://github.com/cms-sw/cmssw/pull/19910
* https://github.com/cms-sw/cmssw/pull/20289 
* https://github.com/cms-sw/cmssw/pull/20308

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

in case this is a transient problem in the harvesting step.
@dmitrijus we have the harvesting step fail occasionally in the cms-bot tests.
Are you aware of any instabilities/issues?

**Comment by slava77** [comment]
@zdemirag 

If I understand correctly, at the BeamHaloSummary level this affects only CSCTightHaloId2015
In the MET filters twiki https://twiki.cern.ch/twiki/bin/viewauth/CMS/MissingETOptionalFiltersRun2
 this filter is not mentioned in the list of recommended ones and there is a note
```
Beam halo filter studies (testing different flags / tightness): apr 27
    Flag_CSCTightHalo2015Filter had been replace by the Flag_globalTightHalo2016Filter for the ICHEP analysis
```

Note  that no differences are observed in CSCTightHaloId2015 in the jenkins tests. This perhaps means that in the 100 events tested in the photonPD tests in cms-bot (wf 136.781) there are no relevant events.
Please suggest a sample where the effect of the fix will be visible.

**Comment by zdemirag** [comment]
Indeed, this filter is no longer recommended (the globaltighthalo filters are the recommended ones for the exact same topology). The reason for the PR is to fix the bugs consistently, so that we don't have to scratch out heads later in the year in case of an emergency to drop the current recommendations and switch back to this filter. JetHT dataset might be good to check, although 100 events seems too low to see effects (because Photon PD is just as relevant). If you would like us to run on a larger sample and post results here, we can do that too.

**Comment by slava77** [comment]
On 8/31/17 9:18 AM, Slava Krutelyov wrote:
> On 8/31/17 8:52 AM, Zeynep Demiragli wrote:
>> Indeed, this filter is no longer recommended (the globaltighthalo
>> filters are the recommended ones for the exact same topology). The
>> reason for the PR is to fix the bugs consistently, so that we don't have
>> to scratch out heads later in the year in case of an emergency to drop
>> the current recommendations and switch back to this filter. 
> 
> OK. Perhaps the backport is not really needed in this case.
> 
>> JetHT
>> dataset might be good to check, although 100 events seems too low to see
>> effects (because Photon PD is just as relevant). If you would like us to
>> run on a larger sample and post results here, we can do that too.
> 
> I would have thought something with muons is more relevant,
> since this algo requires a match between the segments and the calo.
> Do we have triggers [in which PD] selecting on the CSC L1 halo bits?
> 

I'm running on 200 events locally now using 
- 136.782 RunDoubleMuon2017B
- 136.783 RunJetHT2017B
- 136.784 RunMET2017B
- 136.787 RunSingleMu2017B

File: RecoMET/METAlgorithms/src/CSCHaloAlgo.cc [hunk 1]
```diff
@@ -14,7 +14,7 @@ using namespace edm;
 namespace {
   constexpr float c_cm_per_ns  = 29.9792458;
 };
-CSCHaloAlgo::CSCHaloAlgo()
+CSCHaloAlgo::CSCHaloAlgo() : geo_(0), hgeo_(0)
```
Comment (slava77) [inline RecoMET/METAlgorithms/src/CSCHaloAlgo.cc]: please change to nullptr
```diff
-CSCHaloAlgo::CSCHaloAlgo() : geo_(0), hgeo_(0)
+CSCHaloAlgo::CSCHaloAlgo() : geo_(nullptr), hgeo_(nullptr)
```

**Comment by slava77** [comment]
I did not find any differences with 200 event run in the 4 workflows mentioned.

>  If you would like us to run on a larger sample and post results here, we can do that too.

yes, please. This should help with this PR self-documentation and may perhaps serve as an argument for the backport to 92X. 
[since currently I see no effect, and since it's not in a recommended filter, I considered the backport non-essential]

**Comment by slava77** [comment]
@zdemirag 
please clarify on the status of this PR:
- updates in the code
- results from larger samples to confirm the flags are working

**Comment by slava77** [comment]
ping

**Comment by zdemirag** [comment]
I was on vacation, will take care of it today/tomorrow timeline

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20309	50ad780
- use hcal cell positions with a cast to HcalGeometry to get the right cells in phase-1. The code update is in line with other fixes with a similar pattern.
- jenkins tests pass and comparisons show no difference
- based on jenkins tests and some extended tests, it looks like the affected bits in CSCHaloData are rarely modified by the calorimeter-related values. 
    - the question of why it wouldn't fire in the baseline or with the fix is outside of the scope of this PR.

**Comment by davidlange6** [comment]
+1

**Pull request merged**